prov/efa/src/efa_hmem.h

```diff
@@ -65,6 +68,47 @@ struct efa_domain;
 int efa_domain_hmem_validate_p2p_opt(struct efa_domain *efa_domain, enum fi_hmem_iface iface, int p2p_opt);
 int efa_domain_hmem_info_init_all(struct efa_domain *efa_domain);
 
-ssize_t efa_copy_from_hmem_iov(void **desc, char *buff, int buff_size, const struct iovec *hmem_iov, int iov_count);
-ssize_t efa_copy_to_hmem_iov(void **desc, struct iovec *hmem_iov, int iov_count, char *buff, int buff_size);
+static inline int efa_copy_buf_from_hmem(void *desc, void *buff, const void *src, size_t size)
+{
+	uint64_t device = 0;
+	enum fi_hmem_iface iface = FI_HMEM_SYSTEM;
+	void *hmem_data = NULL;
+
+	if (desc) {
+		iface = ((struct efa_mr *)desc)->peer.iface;
+		device = ((struct efa_mr *)desc)->peer.device.reserved;
+		hmem_data = ((struct efa_mr *)desc)->peer.hmem_data;
+	}
+
+	if (FI_HMEM_CUDA == iface && hmem_data) {
+		/* TODO: Fine tune the max data size to switch from gdrcopy to cudaMemcpy */
+		cuda_gdrcopy_from_dev((uint64_t)hmem_data, buff, src, size);
```

Comment:
`cuda_gdrcopy_from_dev` actually returns void.

Reply:
so it will never fail?

Reply:
Actually I kinda ignored this - looking closer at gdrcopy source, there is only 1 unlikely failure mode here https://github.com/NVIDIA/gdrcopy/blob/0d78e6ecc64b97baf79c12ae0c62975ef3f92a9a/src/gdrapi.c#L654

Otherwise it will throw a segfault.